uarray/vectorcall.cxx

```diff
@@ -0,0 +1,235 @@
+#include "vectorcall.h"
+
+#ifdef PYPY_VERSION
+
+/* PyPy doesn't have any support for Vectorcall/FastCall.
+ * These helpers are for translating to PyObject_Call. */
+
+static PyObject * build_arg_tuple(PyObject * const * args, Py_ssize_t nargs) {
+  PyObject * tuple = PyTuple_New(nargs);
+  if (!tuple) {
+    return NULL;
+  }
+
+  for (Py_ssize_t i = 0; i < nargs; ++i) {
+    Py_INCREF(args[i]); /* SET_ITEM steals a reference */
+    PyTuple_SET_ITEM(tuple, i, args[i]);
+  }
+  return tuple;
+}
+
+static PyObject * build_kwarg_dict(
+    PyObject * const * args, PyObject * names, Py_ssize_t nargs) {
+  PyObject * dict = PyDict_New();
+  if (!dict) {
+    return NULL;
+  }
+
+  for (Py_ssize_t i = 0; i < nargs; ++i) {
+    PyObject * key = PyTuple_GET_ITEM(names, i);
+    int success = PyDict_SetItem(dict, key, args[i]);
+    if (success == -1) {
+      Py_DECREF(dict);
+      return NULL;
+    }
+  }
+  return dict;
+}
+#elif (PY_VERSION_HEX < 0x03090000)
+// clang-format off
+
+static int is_method_descr(PyTypeObject* descr_tp) {
+    return (
+        (descr_tp->tp_flags & Q_Py_TPFLAGS_METHOD_DESCRIPTOR) != 0 ||
+        (descr_tp == &PyFunction_Type) ||
+        (descr_tp == &PyMethodDescr_Type));
+}
+
+/* The below code is derivative of CPython source, and is taken because
+_PyObject_GetMethod is not exported from shared objects.
+
+Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010,
+2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021 Python Software Foundation;
+All Rights Reserved
+*/
+
+/* Specialized version of _PyObject_GenericGetAttrWithDict
+   specifically for the LOAD_METHOD opcode.
+
+   Return 1 if a method is found, 0 if it's a regular attribute
+   from __dict__ or something returned by using a descriptor
+   protocol.
+
+   `method` will point to the resolved attribute or NULL.  In the
+   latter case, an error will be set.
+*/
+static int _PyObject_GetMethod(PyObject *obj, PyObject *name, PyObject **method)
```

Comment:
Silver lining: since we have to vendor `_PyObject_GetMethod` for windows' sake, it means we get a speedup in CPython 3.6 as well now. From 390 ns, down to 350 ns.